File: binary_tree/pathSum_DFS.py
'''
leetcode 113. 路径总和2
给定一个二叉树和一个值\ sum sum，请找出所有的根节点到叶子节点的节点值之和等于 sum 的路径
'''

# 
# @param root TreeNode类 
# @param sum int整型 
# @return int整型二维数组
#

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def pathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: List[List[int]]
        """
        self.ret = []
        self.dfs(root, targetSum, [])
        return self.ret

    def dfs(self, root, targetSum, path):
        if not root:
            return
        path.append(root.val)
        targetSum -= root.val
        if root.left is None and root.right is None and targetSum == 0:
            self.ret.append(path)
            return
        self.dfs(root.left, targetSum, copy.deepcopy(path))
        self.dfs(root.right, targetSum, copy.deepcopy(path))

File: binary_tree/pathSum_dfs.cpp
// leetcode 113. pathSum dfs cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) 
    {
        vector<vector<int>> res;
        if(!root)
        {
            return res;
        }
        vector<int> path;
        int remain = targetSum;
        dfs(root, res, path, remain);
        return res;
    }

    void dfs(TreeNode* root, vector<vector<int>>& res, vector<int> path, int remain)
    {
        if (!root)
        {
            return;
        }

        path.push_back(root->val);
        if(!root->left && !root->right)
        {
            if (root->val == remain)
            {
                res.push_back(path);
                return;
            }
            path.pop_back();
            return;
        }
        
        remain -= root->val;
        this->dfs(root->left, res, path, remain);
        this->dfs(root->right, res, path, remain);
    }
};
